Add a way for internal WebKit clients to extract a list of all text and text rects on a webpage

whsieh opened the pull request:
#### 3d7827fa1ae86043d77a65d3fe8479ba5b2ee20c
<pre>
Add a way for internal WebKit clients to extract a list of all text and text rects on a webpage
<a href="https://bugs.webkit.org/show_bug.cgi?id=281795">https://bugs.webkit.org/show_bug.cgi?id=281795</a>
<a href="https://rdar.apple.com/138221056">rdar://138221056</a>

Reviewed by Abrar Rahman Protyasha.

Add experimental WebKit SPI to collect all of the text on a webpage, along with bounding rects in
`WKWebView` coordinates. See below for more details.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractAllTextAndRectsRecursive):
(WebCore::TextExtraction::extractAllTextAndRects):

Implement the main logic here, which uses `TextIterator` to recursively extract visible text content
from each frame on the page.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APITextRun.cpp: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
(API::TextRun::rectInWebView const):
* Source/WebKit/UIProcess/API/APITextRun.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.

Add a new `API::Object` that represents a single piece of visible text, along with a rect in root
view coordinates.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestAllTextWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextRun.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextRun.mm: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
(-[_WKTextRun dealloc]):
(-[_WKTextRun _apiObject]):
(-[_WKTextRun text]):
(-[_WKTextRun rectInWebView]):
(-[_WKTextRun debugDescription]):

Add a new Objective-C object, that wraps the C++ `API::Object` above.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextRunInternal.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAllTextAndRects):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestAllTextAndRects):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:

Drive-by fix: wrap all of these tests in `namespace TestWebKitAPI`.

(-[WKWebView _allTextRuns]):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldReturnString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldFailWhenClosingPage)):
(TestWebKitAPI::TEST(WKWebView, GetContentsOfAllFramesShouldReturnString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldReturnAttributedString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsWithOpticallySizedFontShouldReturnAttributedString)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAccessibilityLabel)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAttributeTypes)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringWithoutNetworkLoads)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringWithSourceApplicationBundleID)):
(TestWebKitAPI::TEST(WKWebView, TextWithWebFontAsAttributedString)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAndCDATASection)):
(TestWebKitAPI::TEST(WKWebView, RequestAllTextRunsWithSubframes)):
(TEST(WKWebView, GetContentsShouldReturnString)): Deleted.
(TEST(WKWebView, GetContentsShouldFailWhenClosingPage)): Deleted.
(TEST(WKWebView, GetContentsOfAllFramesShouldReturnString)): Deleted.
(TEST(WKWebView, GetContentsShouldReturnAttributedString)): Deleted.
(TEST(WKWebView, GetContentsWithOpticallySizedFontShouldReturnAttributedString)): Deleted.
(TEST(WKWebView, AttributedStringAccessibilityLabel)): Deleted.
(TEST(WKWebView, AttributedStringAttributeTypes)): Deleted.
(TEST(WKWebView, AttributedStringFromTable)): Deleted.
(TEST(WKWebView, AttributedStringWithLinksInTableCell)): Deleted.
(TEST(WKWebView, AttributedStringFromList)): Deleted.
(TEST(WKWebView, AttributedStringWithoutNetworkLoads)): Deleted.
(TEST(WKWebView, AttributedStringWithSourceApplicationBundleID)): Deleted.
(TEST(WKWebView, TextWithWebFontAsAttributedString)): Deleted.
(TEST(WKWebView, AttributedStringAndCDATASection)): Deleted.
(TEST(WKWebView, AttributedStringIncludesUserSelectNoneContent)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/nested-frames.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/subframes.html: Added.

Add a basic API test, verifying that the new SPI method finds text inside of nested subframes and
passes back rects in the web view&apos;s coordinate space.

Canonical link: <a href="https://commits.webkit.org/285706@main">https://commits.webkit.org/285706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/275e99429e54657d62d2466652bcbdf583d559ae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52934 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77774 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24743 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75620 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62065 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57762 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16195 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76572 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47854 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63245 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38172 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44467 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20730 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23074 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66271 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21080 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79391 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/309 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66156 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/962 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63257 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65434 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16184 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9294 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7468 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/784 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3527 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/814 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/800 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/820 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Modules/OSX_Private.modulemap:
```diff
@@ -2014,6 +2014,11 @@ framework module WebKit_Private [system] {
     export *
   }
 
+  explicit module _WKTextRun {
```
I think we should stop trying to add any new explicit modules to the private module, since it’s not designed to have a sub module for each type, and also causes greater confusion since we have both the general internal module and a bunch of random sub modules

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextRun.h:
```diff
@@ -0,0 +1,39 @@
+
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+@interface _WKTextRun : NSObject
+
```
This type should probably be annotated as sendable in for any potential Swift clients

aprotyas commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3178,6 +3179,19 @@ - (void)_addAppHighlightInNewGroup:(BOOL)newGroup originatedInApp:(BOOL)originat
+        if (!strongSelf)
+            return completion(@[ ]);
+
+        completion(createNSArray(textRuns, [](auto& textRun) {
```
I think you can `std::forward(textRuns)` so if you're passing in an rvalue reference it can get moved along.

whsieh replied:
Hm... that gave me a compile-time error:

```
OpenSource/Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:3301:34: error: no matching function for call to 'forward'
 3301 |         completion(createNSArray(std::forward(textRuns), [](auto& textRun) mutable {
```

I'll adjust this to instead `WTFMove(textRuns)` when passing into `createNSArray`, to take advantage of move semantics here.

cdumez commented on Source/WebKit/UIProcess/API/APITextRun.h:
```diff
@@ -0,0 +1,60 @@
+        return adoptRef(*new TextRun(page, WTFMove(string), WTFMove(rect)));
+    }
+
+    explicit TextRun(WebKit::WebPageProxy& page, WTF::String&& string, WebCore::FloatRect&& rect)
```
We should not have both a public constructor and a create() function. This constructor should likely be private.

whsieh replied:
Good catch, made it private.

whsieh committed:
Add a way for internal WebKit clients to extract a list of all text and text rects on a webpage
https://bugs.webkit.org/show_bug.cgi?id=281795
rdar://138221056

Reviewed by Abrar Rahman Protyasha.

Add experimental WebKit SPI to collect all of the text on a webpage, along with bounding rects in
`WKWebView` coordinates. See below for more details.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractAllTextAndRectsRecursive):
(WebCore::TextExtraction::extractAllTextAndRects):

Implement the main logic here, which uses `TextIterator` to recursively extract visible text content
from each frame on the page.

* Source/WebCore/page/text-extraction/TextExtraction.h:
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APITextRun.cpp: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
(API::TextRun::rectInWebView const):
* Source/WebKit/UIProcess/API/APITextRun.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.

Add a new `API::Object` that represents a single piece of visible text, along with a rect in root
view coordinates.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestAllTextWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextRun.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextRun.mm: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
(-[_WKTextRun dealloc]):
(-[_WKTextRun _apiObject]):
(-[_WKTextRun text]):
(-[_WKTextRun rectInWebView]):
(-[_WKTextRun debugDescription]):

Add a new Objective-C object, that wraps the C++ `API::Object` above.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextRunInternal.h: Copied from Source/WebCore/page/text-extraction/TextExtraction.h.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAllTextAndRects):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestAllTextAndRects):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:

Drive-by fix: wrap all of these tests in `namespace TestWebKitAPI`.

(-[WKWebView _allTextRuns]):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldReturnString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldFailWhenClosingPage)):
(TestWebKitAPI::TEST(WKWebView, GetContentsOfAllFramesShouldReturnString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsShouldReturnAttributedString)):
(TestWebKitAPI::TEST(WKWebView, GetContentsWithOpticallySizedFontShouldReturnAttributedString)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAccessibilityLabel)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAttributeTypes)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringWithoutNetworkLoads)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringWithSourceApplicationBundleID)):
(TestWebKitAPI::TEST(WKWebView, TextWithWebFontAsAttributedString)):
(TestWebKitAPI::TEST(WKWebView, AttributedStringAndCDATASection)):
(TestWebKitAPI::TEST(WKWebView, RequestAllTextRunsWithSubframes)):
(TEST(WKWebView, GetContentsShouldReturnString)): Deleted.
(TEST(WKWebView, GetContentsShouldFailWhenClosingPage)): Deleted.
(TEST(WKWebView, GetContentsOfAllFramesShouldReturnString)): Deleted.
(TEST(WKWebView, GetContentsShouldReturnAttributedString)): Deleted.
(TEST(WKWebView, GetContentsWithOpticallySizedFontShouldReturnAttributedString)): Deleted.
(TEST(WKWebView, AttributedStringAccessibilityLabel)): Deleted.
(TEST(WKWebView, AttributedStringAttributeTypes)): Deleted.
(TEST(WKWebView, AttributedStringFromTable)): Deleted.
(TEST(WKWebView, AttributedStringWithLinksInTableCell)): Deleted.
(TEST(WKWebView, AttributedStringFromList)): Deleted.
(TEST(WKWebView, AttributedStringWithoutNetworkLoads)): Deleted.
(TEST(WKWebView, AttributedStringWithSourceApplicationBundleID)): Deleted.
(TEST(WKWebView, TextWithWebFontAsAttributedString)): Deleted.
(TEST(WKWebView, AttributedStringAndCDATASection)): Deleted.
(TEST(WKWebView, AttributedStringIncludesUserSelectNoneContent)): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/nested-frames.html:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/subframes.html: Added.

Add a basic API test, verifying that the new SPI method finds text inside of nested subframes and
passes back rects in the web view's coordinate space.

Canonical link: https://commits.webkit.org/285706@main
